### PR TITLE
feat: sponsored swap and deposits

### PR DESCRIPTION
**Pull request by bmzig**
This PR adds an extra signatureOwner and signature argument to the functions which use `permit` and `receiveWithAuthorization` so that we no longer have a dependency for the depositor to call these functions. This should mean that as long as the depositor signed off on the input deposit/swapAndBridge arguments for the `DepositData` and `SwapAndDepositData` structs respectively, then anybody can call these functions and execute the swap.

Most of the diff for this PR are tests.

### CODE REVIEW

File: test/evm/foundry/local/SpokePoolPeriphery.t.sol [hunk 1]
```diff
@@ -413,29 +854,146 @@ contract SpokePoolPeripheryTest is Test {
...
+        spokePoolPeriphery.swapAndBridgeWithPermit2(
+            depositor, // signatureOwner
+            swapAndDepositData,
+            permit,
+            signature
+        );
+    }
 
-        vm.stopPrank();
+    function testPermit2SwapAndBridgeValidWitness(address rando) public {
```
Comment (bmzig) [inline test/evm/foundry/local/SpokePoolPeriphery.t.sol:928]:
```suggestion
    function testPermit2SwapAndBridgeInvalidWitness(address rando) public {
```

File: contracts/SpokePoolV3Periphery.sol [hunk 2]
```diff
@@ -491,6 +536,14 @@ contract SpokePoolV3Periphery is SpokePoolV3PeripheryInterface, Lockable, MultiC
             s
         );
 
+        // Verify that the signatureOwner signed the input depositData.
+        if (
```
Comment (nicholaspai) [inline contracts/SpokePoolV3Periphery.sol:540]: Is this code block the same for all gasless flows? Can we refactor into internal func?

**Review by nicholaspai — APPROVED** [review]
Looks good! just one suggestion

File: contracts/SpokePoolV3Periphery.sol (hunk 2 above)
Comment (bmzig) [reply contracts/SpokePoolV3Periphery.sol]: It is the same. Factored out here: https://github.com/across-protocol/contracts/pull/790/commits/dfa6f89bcbbfc3fea759b06ed37cd600b0a64668

**Pull request merged**